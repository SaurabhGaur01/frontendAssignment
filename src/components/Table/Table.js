import React from 'react'; 
import { connect } from 'react-redux';
import { compose } from 'redux';
import PropTypes from 'prop-types';
import calculateTimeSeries from '../../utils/calculateTimeSeries';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';

/* 
    Using React material for showing data in Tabular format,
    this component will be pick values from store and do the manipulation...
*/
const TableComponent = ({ 
    riskLevel, 
    conesData, 
    initialSum, 
    monthlySum, 
    years, 
    fee 
}) => {
    // Return null when conesData is an empty []
    if (conesData.length > 0) {
        const cone = conesData.filter(cone => cone.riskLevel == riskLevel)[0];
        const{ median, upper95, lower05 } = calculateTimeSeries({
            mu: cone ? cone.mu : 0,
            sigma: cone ? cone.sigma: 0,
            years,
            initialSum,
            monthlySum,
            fee
        });

        const months = median.map((v, idx) => idx);
        const dataGood = upper95.map(v => v.y);
        const dataMedian = median.map(v => v.y);
        const dataBad = lower05.map(v => v.y);

        return (
            <TableContainer>
                <Table size="small">
                    <TableHead>
                        <TableRow>
                            <TableCell>Month</TableCell>
                            <TableCell>Good</TableCell>
                            <TableCell>Median</TableCell>
                            <TableCell>Bad</TableCell>
                        </TableRow>
                    </TableHead>
                    <TableBody>
                        { months.map((entry, idx) => (
                            <TableRow key={idx}>
                                <TableCell>{entry}</TableCell>
                                <TableCell>{dataGood[idx]}</TableCell>
                                <TableCell>{dataMedian[idx]}</TableCell>
                                <TableCell>{dataBad[idx]}</TableCell>
                            </TableRow>
                        ))}
                    </TableBody>
                </Table>
            </TableContainer>
        );
    }
    return null;
}

TableComponent.propTypes = {
    riskLevel: PropTypes.number.isRequired,
    conesData: PropTypes.arrayOf(PropTypes.shape({
        riskLevel: PropTypes.number.isRequired,
        mu: PropTypes.number.isRequired,
        sigma: PropTypes.number.isRequired,
    })).isRequired,
    initialSum: PropTypes.number.isRequired,
    years: PropTypes.number.isRequired,
    monthlySum: PropTypes.number.isRequired,
    fee: PropTypes.number.isRequired,
};

// Getting below keys from redux store
export const mapStateToProps = state => ({
    riskLevel: state.riskLevel,
    conesData: state.conesData,
    initialSum: state.initialSum,
    years: state.years,
    monthlySum: state.monthlySum,
    fee: state.fee,
});

const hocChain = compose(
    connect(mapStateToProps),
);

export { TableComponent as TestableTableComponent };

export default hocChain(TableComponent);
