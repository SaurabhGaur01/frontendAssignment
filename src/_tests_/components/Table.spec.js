import React from 'react';
import { shallow } from 'enzyme';
import { TestableTableComponent, mapStateToProps } from '../../components/Table/Table';

jest.mock('../../utils/calculateTimeSeries', () => () => ({
    median: [{ x: 1.2, y: 2.5 }],
    upper95: [{ x: 1.2, y: 2.5 }],
    lower05: [{ x: 1.2, y: 2.5 }],
}))

describe('Table Component', () => {
    let props;

    beforeEach(() =>{
        props = {
            riskLevel: 3,
            initialSum: 5000,
            years: 10,
            monthlySum: 500,
            conesData: [{
                riskLevel: 3,
                mu: 0.024,
                sigma: 0.031,
            }]
        };
    });

    it('should render correctly', () => {
        const renderedModule = shallow(<TestableTableComponent {...props} />);
        expect(renderedModule).toMatchSnapshot();
    });

    it('should render correctly when conesData is empty', () => {
        props.conesData = [];
        const renderedModule = shallow(<TestableTableComponent {...props} />);
        expect(renderedModule).toMatchSnapshot();
    });

    describe('mapStateToProps', () => {
        const riskLevelMock = 'mockValue';
        const conesDataMock = [{ mockKey: 'mockValue'}];
        const initialSumMock = 'mockValue';
        const yearsMock = 'mockValue';
        const monthlySumMock = 'mockValue';
    
        const mockedState = {
            riskLevel: riskLevelMock,
            conesData: conesDataMock,
            initialSum: initialSumMock,
            years: yearsMock,
            monthlySum: monthlySumMock,
        };

        it('should return the module state from the redux store', () => {
            const mockPropsFromRedux = mapStateToProps(mockedState);
            expect(mockPropsFromRedux).toEqual(
                {
                    riskLevel: riskLevelMock,
                    conesData: conesDataMock,
                    initialSum: initialSumMock,
                    years: yearsMock,
                    monthlySum: monthlySumMock,
                }
            )
        })
    })
})