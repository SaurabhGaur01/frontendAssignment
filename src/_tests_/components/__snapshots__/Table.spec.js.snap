// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component should render correctly 1`] = `
<WithStyles(ForwardRef(TableContainer))>
  <WithStyles(ForwardRef(Table))
    size="small"
  >
    <WithStyles(ForwardRef(TableHead))>
      <WithStyles(ForwardRef(TableRow))>
        <WithStyles(ForwardRef(TableCell))>
          Month
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          Good
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          Median
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          Bad
        </WithStyles(ForwardRef(TableCell))>
      </WithStyles(ForwardRef(TableRow))>
    </WithStyles(ForwardRef(TableHead))>
    <WithStyles(ForwardRef(TableBody))>
      <WithStyles(ForwardRef(TableRow))
        key="0"
      >
        <WithStyles(ForwardRef(TableCell))>
          0
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          2.5
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          2.5
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          2.5
        </WithStyles(ForwardRef(TableCell))>
      </WithStyles(ForwardRef(TableRow))>
    </WithStyles(ForwardRef(TableBody))>
  </WithStyles(ForwardRef(Table))>
</WithStyles(ForwardRef(TableContainer))>
`;

exports[`Table Component should render correctly when conesData is empty 1`] = `""`;
